---
title: "Variant calling RNAseq"
author: Tommaso Barberis
output: 
  md_document:
    variant: "markdown_github"
knit: (function(inputFile, encoding) { rmarkdown::render(inputFile, encoding = encoding, output_file = "Variant_calling_RNAseq.md" )})
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Variant calling from RNAseq data of *D. suzukii*

The *variants.table* file is obtained during a **variant calling** experiment using *HaplotypeCaller* from *GATK*, on RNAseq from *D. suzukii*, by M. **Tabourin**. The following code is inspired from her work too.

```{r}
library(tidyverse)

data <- read.table("variants_PLI.table",header = T, sep = "\t")
```

Column description of the *.table* file (Van der Auwera & O'Connor (2020)):

-   **`CHROM`**: name of the chromosome where the variant is found;
-   **`POS`**: genomic position on the chromosome;
-   **`TYPE`**: type of the variant (**`SNP`**, **`INDEL`** or **`MIXED`**);
-   **`HET`**: count of heterozygous genotypes;
-   **`NSAMPLES`**: number of sample (here, we have 10 samples);
-   **`NCALLED`**: number of called samples;
-   **`.GT`** columns: genotype of the sample at the **`POS`** position;
-   **`.AD`** columns: Allele depth, the number of reads that support each of the reported alleles.

## Variant calling

So, now we will try to compute the number of variant for every sample:

```{r}
samples <- c("Ce1_1.AD", "Ce1_2.AD", "Ce3_1.AD", "Ce3_2.AD", "Fr1_1.AD", "Fr1_2.AD", "Fr3_1.AD", "Fr3_2.AD", "G0_1.AD", "G0_2.AD")
results <- list()

for (sample in samples) {

  SNP_number <- 0

  for (variant in data[[sample]]) {

    i = dim(str_split(variant, ",", simplify = TRUE))[2]
    pass = "False"

    while (pass == "False") {
      if ((str_split(variant, ",", simplify = TRUE)[i]) > 0){
        SNP_number = SNP_number + 1
        pass = "True"
      }

      i = i - 1

      if (i == 1) {
        pass = "True"
      }
    }
  }

  results$samples[[sample]] <- SNP_number
}
```

**Barplot** showing number of variants for every sample:

```{r}
library(grid)
library(gridBase)

variant_barplot <- barplot(unlist(results), names.arg="", col=c("palegreen4","palegreen4","palegreen2", "palegreen2", "royalblue3", "royalblue3",
                "royalblue1", "royalblue1", "orange2", "orange2"))

## Use grid to add the labels    
vps <- baseViewports()
pushViewport(vps$inner, vps$figure, vps$plot)

grid.text(samples,
    x = unit(variant_barplot, "native"), y=unit(-1, "lines"),
    just="right", rot=50)

popViewport(3)
```

## Count number of heterozygous positions

```{r}
samples <- c("Ce1_1.PL", "Ce1_2.PL", "Ce3_1.PL", "Ce3_2.PL", "Fr1_1.PL", "Fr1_2.PL", "Fr3_1.PL", "Fr3_2.PL", "G0_1.PL", "G0_2.PL")

for (sample in samples) {

  HET_pos_number <- 0
  
  for (variant in data[[sample]]) {
    str_split(variant, ",", simplify = TRUE)
    
    # if (dim(str_split(variant, ",", simplify = TRUE))[2] == 3) {
    #   if ((str_split(variant, ",", simplify = TRUE)[2]) == 0 && (str_split(variant, ",", simplify = TRUE)[1]) != 0 ){
    #     HET_pos_number = HET_pos_number + 1
    #   }
    # }
    # else if (dim(str_split(variant, ",", simplify = TRUE))[2] == 6) {
    #   if ((str_split(variant, ",", simplify = TRUE)[1]) != 0 && (str_split(variant, ",", simplify = TRUE)[3]) != 0 && (str_split(variant, ",", simplify = TRUE)[6]) != 0){
    #     HET_pos_number = HET_pos_number + 1
    #   } 
    # }
    # else if (dim(str_split(variant, ",", simplify = TRUE))[2] == 10) {
    #   if ((str_split(variant, ",", simplify = TRUE)[1]) != 0 && (str_split(variant, ",", simplify = TRUE)[3]) != 0 && (str_split(variant, ",", simplify = TRUE)[6]) != 0 && (str_split(variant, ",", simplify = TRUE)[10]) != 0){
    #     HET_pos_number = HET_pos_number + 1
    #   }
    # }
    # else if (dim(str_split(variant, ",", simplify = TRUE))[2] == 15) {
    #   if ((str_split(variant, ",", simplify = TRUE)[1]) != 0 && (str_split(vr, ",", simplify = TRUE)[3]) != 0 && (str_split(variant, ",", simplify = TRUE)[6]) != 0 && (str_split(variant, ",", simplify = TRUE)[10]) != 0 && (str_split(variant, ",", simplify = TRUE)[15]) != 0){
    #   HET_pos_number = HET_pos_number + 1
    #   }
    # }
    # else if (dim(str_split(variant, ",", simplify = TRUE))[2] == 21) {
    #   if ((str_split(variant, ",", simplify = TRUE)[1]) != 0 && (str_split(variant, ",", simplify = TRUE)[3]) != 0 && (str_split(variant, ",", simplify = TRUE)[6]) != 0 && (str_split(variant, ",", simplify = TRUE)[10]) != 0 && (str_split(variant, ",", simplify = TRUE)[15]) != 0 && (str_split(variant, ",", simplify = TRUE)[21]) != 0){
    #     HET_pos_number = HET_pos_number + 1
    #   }
    # }
    # else if (dim(str_split(variant, ",", simplify = TRUE))[2] == 28) {
    #   if ((str_split(variant, ",", simplify = TRUE)[1]) != 0 && (str_split(variant, ",", simplify = TRUE)[3]) != 0 && (str_split(variant, ",", simplify = TRUE)[6]) != 0 && (str_split(variant, ",", simplify = TRUE)[10]) != 0 && (str_split(variant, ",", simplify = TRUE)[15]) != 0 && (str_split(variant, ",", simplify = TRUE)[21]) != 0 && (str_split(variant, ",", simplify = TRUE)[28]) != 0){
    #     HET_pos_number = HET_pos_number + 1
      }
    }
  }
}
```

```{r}
dim(data)
```
